# <center>Lista 02 - Aprendizado de Máquina </center>

**Aluno(a):** Marianna de Pinho Severo <br>
**Matrícula:** 374856 <br>
**Professor(a):** Regis Pires

## 1) Crie um documento do Jupyter Notebook para criar um modelo preciso de classificação para o Iris Dataset.
O dataset foi baixado de [Iris dataset](https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data) e foi salvo em no arquivo *iris.csv*, em que a primeira linha indica os nomes das colunas.

## 2) Baixe e carregue os dados.
Uma vez que os dados já foram baixados, agora iremos carregá-los. Para isso, primeiramente importaremos todas as bibliotecas que serão utilizadas para essa e as outras atividades da lista.

### Passo 01: Importar bibliotecas

In [1]:
%matplotlib inline
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import metrics #biblioteca utilizada para extrair as métricas de desempenho do algoritmo de aprendizagem.
from sklearn import tree #biblioteca utilizada para emprego do classificador de árvore de decisão.

### Passo 02: Criar função para carregamento dos dados
DATASET_PATH é um define que indica onde o conjunto de dados está armazenado no computador. Nesse caso, está em uma pasta, chamada *dataset*, localizada dentro da pasta em que este notebook se encontra.

In [2]:
DATASET_PATH = "dataset/"

A função *load_flowers_data* é responsável por abrir o arquivo que contém o conjunto de dados, chamado *iris.csv*, e retornar seus valores no formato DataFrame, que é uma estrutura de dados de duas dimensões, cujos eixos são rotulados.

In [5]:
def load_flowers_data(dataset_path = DATASET_PATH):
    csv_path = os.path.join(dataset_path, "iris.csv")
    return pd.read_csv(csv_path)

### Passo 03: Carregar os dados
Por fim, carregaremos o conjunto de dados dentro da variável *flowers*, que será utilizada para a realização das próximas operações.

In [4]:
flowers = load_flowers_data()

## 3) Faça uma breve análise dos dados.
Nessa etapa, utilizaremos algumas das funções apresentadas no Capítulo 02 do livro **Hands-On Machine Learning with Scikit-Learn and TensorFlow** para uma breve visualização e análise do conjunto de dados.

### Passo 01: Método head()
O conjunto de dados (*dataset*) está armazenado na variável *flowers*. Com o método *head()*, apresentamos as cinco primeiras linhas desse conjunto, em formato de tabela. Nela, é possível ver que o conjunto de dados possui quatro atributos, que são o **comprimento da sépala**, a **largura da sépala**, o **comprimento da pétala** e a **largura da pétala**, todos expressos em centímetros. Além disso, a partir desses atributos, os dados podem ser classificados em três classes: **Iris-setosa**, **Iris-versicolor** e **Iris-virginica**, apresentadas na última coluna.

In [6]:
flowers.head()

sepal_length  sepal_width  petal_length  petal_width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

### Passo 02: Método info()
Utilizando o método *info()*, obtemos uma descrição rápida dos dados. É possível observar que o *dataset* possui 150 linhas, sendo elas numeradas de 0 a 149; que ele possui cinco colunas, em que as quatro primeiras são para os atributos e a última é para o *label*, ou classe, dos dados; que os atributos são do tipo *float* e a classe é do tipo objeto; também é possível observar que todas as linhas estão preenchidas, o que significa que não há linhas nem colunas sem valores.

In [7]:
flowers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal_length    150 non-null float64
sepal_width     150 non-null float64
petal_length    150 non-null float64
petal_width     150 non-null float64
class           150 non-null object
dtypes: float64(4), object(1)
memory usage: 5.9+ KB


### Passo 03: Método value_counts()
Utilizando o método *value_counts()*, é possível observarmos quais classes existem no conjunto de dados e quantas linhas correspondem a cada uma delas.

In [8]:
flowers["class"].value_counts()

Iris-virginica     50
Iris-versicolor    50
Iris-setosa        50
Name: class, dtype: int64

### Passo 04: Método describe()
Com o método *describe()*, uma tabela com um resumo de todos os dados do *dataset* é mostrada. A linha **std** indica o *desvio padrão* dos dados. Já, as linhas **25%**, **50%** e **75%** indicam os percentis, que são valores abaixo dos quais uma determinada porcentagem de observações em um grupo recai.

In [9]:
flowers.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

De acordo com a tabela, é possível observar que: as flores possuem um comprimento médio de sépala e de pétala de, aproximadamente, 5.84 cm e 3.758 cm, respectivamente; que elas possuem uma largura média de sépala e de pétala de, aproximadamente 3.054 cm e 1.198 cm, respectivamente; que 25% das flores possuem comprimento de sépala menor do que 5.1 cm; que 50% das flores possuem comprimento de pétala menor do que 4.35 cm; que 75% das flores possuem largura de sépala menor do que 3.3 cm; que a largura máxima de pétala é de 2.5 cm; entre outras características.

### Passo 05: Método hist()
Utilizando o método *hist()*, podemos observar os histogramas do conjunto de dados, em que cada gráfico representa um atributo, o eixo horizontal de cada gráfico representa o valor que aquele atributo possui e o eixo vertical representa quantas instâncias dos dados possuem aquele valor.

In [1]:
flowers.hist(bins=50, figsize=(10,10))
plt.show()

NameError: name 'flowers' is not defined

É possível observar, de acordo com a distribuição dos dados, que a maior variedade de comprimentos e larguras ocorre para valores acima da média.

### Passo 06: Método tail() 
Esse método mostra as últimas cinco linhas do conjunto de dados.

In [10]:
flowers.tail()

sepal_length  sepal_width  petal_length  petal_width           class
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

### Passo 07: Método corr()

Esse método calcula a correlação entre os atributos do conjunto de dados. Por padrão, o método de correlação empregado é o *Pearson correlation coefficient*, que é uma medida linear de correlação entre duas variáveis. Quanto mais próxima de 1, mais forte é a correlação em um sentido positivo; quanto mais próxima de -1, mais forte ela é em um sentido negativo; e, por último, quanto mais próxima de 0, menor é a correlação entre as duas variáveis consideradas.

In [11]:
flowers.corr()

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.109369      0.871754     0.817954
sepal_width      -0.109369     1.000000     -0.420516    -0.356544
petal_length      0.871754    -0.420516      1.000000     0.962757
petal_width       0.817954    -0.356544      0.962757     1.000000

Podemos observar que os atributos possuem correlação máxima e positiva com eles mesmos. O comprimento da sépala possui uma elevada correlação positiva com o comprimento e largura da pétala, ou seja, quanto maior forem esses valores, maior será o comprimento da sépala e vice-versa. Por outro lado, ele possui uma pequena correlação negativa com a largura da sépala, ou seja, quanto maior for esse valor, menor será o comprimento da sépala, mas essa variação ocorrerá mais lentamente. Também é possível observar que o comprimento da pétala tem uma elevada correlação positiva com sua largura.

## 4) Separe parte dos dados para treino (70%) e parte para teste (30%). 

### Passo 01: Extrair os valores
Primeiramente, extrairemos os valores contidos na estrutura *flowers*, que é do tipo *DataFrame*. Para isso, utilizaremos o atributo *values*, que retorna uma estrutura NumPy que contém os dados existentes na estrutura *flowers*,  retirando-se os nomes dos eixos, e a salvaremos na estrutura  **dataset**.

In [13]:
dataset = flowers.values

### Passo 02: Tornar os dados aleatórios
Os dados do conjunto de dados estão ordenados de acordo com suas classes, ou seja, as primeiras cinquenta linhas referem-se a uma classe, as outras cinquenta, à outra classe, e as últimas cinquenta linhas referem-se à classe restante. Dessa forma, como separaremos esses dados em um conjunto de treino e um de teste, é importante que o conjunto de treino tenha dados de todas as classes. Portanto, reorganizando as linhas de forma aleatória, garantiremos que esse conjunto tenha acesso aos dados desejados. Para isso, utilizaremos os métodos **random.seed()** e **random.shuffle()**.

In [14]:
np.random.seed(1)
np.random.shuffle(dataset)

### Passo 03: Separar *features* e *lables*
Agora, salvaremos na estrutura X todas as características que serão utilizadas pelo algoritmo de aprendizado para realizar classificação. Elas podem ser entendidas como a entrada do sistema. Então, salvaremos na estrutura Y todos os lables, que, neste caso, são classes, do conjunto de dados. Eles podem ser entendidos como a saída do sistema e correspondem aos valores que o algoritmo dever prever, ou classificar, de acordo com a entrada recebida.

In [17]:
X = dataset[:,:4]
y = dataset[:,4]

### Passo 04: Criação dos conjuntos de treino e de teste
Por fim, criaremos os conjuntos de treino e teste, salvando, nas estruturas **X_train** e **y_train**, 70% dos dados e, nas estruturas **X_test** e **y_test**, os 30% restantes. Nas estruturas que começam com *X_* estão os dados de entrada e nas que começam com *y_*, os de saída.

In [19]:
n_train = int(round(len(y) * 0.70))
X_train = X[:n_train,:]
y_train = y[:n_train]
X_test = X[n_train:,:]
y_test = y[n_train:]

## 5) Crie um modelo de aprendizado de máquina usando algum algoritmo de classificação sobre os dados de treino.
O algoritmo de aprendizado escolhido para se realizar a classificação dos dados do [Iris dataset](https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data) é baseado em **Árvore de Decisão**. Seu nome é *DecisionTreeClassifier()* e ele está implementado na biblioteca *Scikit-Learn*.

Uma árvore de decisão é composta por **nós**, **atributos** e **folhas**. Atributos são propriedades a serem testadas em determinados nós da árvore, de forma que um de seus possíveis valores de resposta é escolhido, seguindo-se para os nós seguintes, até que se encontre um nó folha. Este, por sua vez, é a decisão final do algoritmo para o conjunto de dados. No caso de classificadores, o nó folha é a classe escolhida para os dados de entrada. Esse algoritmo utiliza os conceitos de **Entropia** e **Ganho de Informação** para determinar a ordem em que os atributos serão testados na árvore, de maneira a obter a menor entropia e o maior ganho possíveis.

In [16]:
model_decisiontree = tree.DecisionTreeClassifier()

## 6) Determine a acurácia do modelo sobre os dados de treino.
Para treinar o modelo adotado, utilizamos o método *.fit()*, que recebe como argumentos os dados de treino de entrada e de saída.

In [17]:
trained_model = model_decisiontree.fit(X_train,y_train)

Então, para calcularmos a acurácia do modelo sobre os dados de treino, podemos utilizar o método *metrics.accuracy_score()*. Para isso, realizaremos a predição de *y_train* a partir de *X_train*, empregando o método *predict()*, que realiza a predição sobre os dados utilizando o modelo treinado. Dessa forma, temos:

In [37]:
train_pred = trained_model.predict(X_train)
train_accuracy = metrics.accuracy_score(y_train, train_pred)
train_accuracy

1.0

Podemos observar que a acurácia para os dados de treino é máxima, o que significa que o algoritmo classifica corretamente as flores de acordo com as características passadas. A seguir, utilizamos o método *classification_report()* para produzir um pequeno relatório de comparação entre as saídas esperadas (y_train) e as obtidas pelo algoritmo (train_pred). Uma pequena explicação sobre o que significa a saída desse relatório pode ser encontrada em [Classification Report](https://www.scikit-yb.org/en/latest/api/classifier/classification_report.html).

In [38]:
print(metrics.classification_report(y_train, train_pred))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        33
Iris-versicolor       1.00      1.00      1.00        35
 Iris-virginica       1.00      1.00      1.00        37

      micro avg       1.00      1.00      1.00       105
      macro avg       1.00      1.00      1.00       105
   weighted avg       1.00      1.00      1.00       105



Podemos observar que os dados estão bem distribuídos, conforme indica a coluna **support**, pois cada classe está presente, aproximadamente, a mesma quantidade de vezes no conjunto de dados.

## 7) Determine a acurácia do modelo sobre os dados de teste.
Para calcularmos a acurácia sobre os dados de teste, precisamos fazer a predição para esses dados. Isso é feito com o método *predict()*, a partir do modelo adotado.

In [30]:
pred_decisiontree = trained_model.predict(X_test)

Nas listas apresentadas abaixo, podemos observar tanto a saída (*y_test*) esperada para os dados de entrada, como a saída obtida pela predição (*pred_decisiontree*), respectivamente.

In [32]:
print (y_test)
print(pred_decisiontree)

['Iris-setosa' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-virginica' 'Iris-virginica' 'Iris-virginica' 'Iris-setosa'
 'Iris-setosa' 'Iris-versicolor' 'Iris-setosa' 'Iris-virginica'
 'Iris-setosa' 'Iris-virginica' 'Iris-virginica' 'Iris-setosa'
 'Iris-virginica' 'Iris-setosa' 'Iris-versicolor' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-setosa' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-virginica' 'Iris-setosa' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-versicolor'
 'Iris-virginica' 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica'
 'Iris-setosa']
['Iris-setosa' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-virginica' 'Iris-virginica' 'Iris-virginica' 'Iris-setosa'
 'Iris-setosa' 'Iris-virginica' 'Iris-setosa' 'Iris-virginica'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Ir

Abaixo, podemos observar o relatório gerado para os dados de teste. Observando a coluna *support*, vemos que as classes aparecem, aproximadamente, na mesma quantidade no conjunto de dados. Além disso, vemos que a classe Iris-setosa recebeu pontuação máxima para as três primeiras métricas das colunas, o que indica que o algoritmo acertou todas as vezes que determinou que uma flor do conjunto de teste pertencia a essa classe. Entretanto, o algoritmo encontrou maior dificuldade em diferenciar flores pertencentes às classes Iris-versicolor e Iris-virginica.

In [21]:
print(metrics.classification_report(y_test, pred_decisiontree))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        17
Iris-versicolor       0.76      0.87      0.81        15
 Iris-virginica       0.82      0.69      0.75        13

      micro avg       0.87      0.87      0.87        45
      macro avg       0.86      0.85      0.85        45
   weighted avg       0.87      0.87      0.87        45



Por fim, a acurácia para o conjunto de testes é de **0.8666666666666667**, conforme mostra saída abaixo.

In [29]:
test_accuracy = metrics.accuracy_score(y_test, pred_decisiontree)
print(test_accuracy)

0.8666666666666667
